lib/iris/tests/unit/fileformats/pp/test_PPDataProxy.py

```diff
@@ -48,5 +52,122 @@ def test_lnpack_raw(self):
         self.assertEqual(proxy.lbpack.n4, lbpack // 1000 % 10)
 
 
+class SliceTranslator():
+    """
+    Class to translate an array-indexing expression into a tuple of keys.
+
+    An instance just returns the argument of its __getitem__ call.
+
+    """
+    def __getitem__(self, keys):
+        return keys
+
+
+# A multidimensional-indexable object that returns its index keys, so we can
+# use multidimensional-indexing notation to specify a slicing expression.
+Slices = SliceTranslator()
+
+
+class Test__getitem__slicing(tests.IrisTest):
+    def _check_slicing(self, test_shape, indices, result_shape,
+                       data_was_fetched=True):
+        # Check behaviour of the getitem call with specific slicings.
+        # Especially: check cases where a fetch does *not* read from the file.
+        # This is necessary because, since Dask 2.0, the "from_array" function
+        # takes a zero-length slice of its array argument, to capture array
+        # metadata, and in those cases we want to avoid file access.
+        test_dtype = np.dtype(np.float32)
+        proxy = PPDataProxy(shape=test_shape, src_dtype=test_dtype,
+                            path=None, offset=None, data_len=None,
+                            lbpack=0,  # Note: a 'real' value is needed.
+                            boundary_packing=None, mdi=None)
+
+        # Mock out the file-open call, to see if the file would be read.
+        if six.PY2:
+            builtin_open_func_name = '__builtin__.open'
+        else:
+            builtin_open_func_name = 'builtins.open'
+        mock_fileopen = self.patch(builtin_open_func_name)
+
+        # Also mock out the 'databytes_to_shaped_array' call, to fake minimal
+        # operation in the cases where file-open *does* get called.
+        fake_data = np.zeros(test_shape, dtype=test_dtype)
+        self.patch('iris.fileformats.pp._data_bytes_to_shaped_array',
+                   mock.MagicMock(return_value=fake_data))
+
+        # Test the requested indexing operation.
+        result = proxy.__getitem__(indices)
+
+        # Check the behaviour and results were as expected.
+        self.assertEqual(mock_fileopen.called, data_was_fetched)
+        self.assertIsInstance(result, np.ndarray)
+        self.assertEqual(result.dtype, test_dtype)
+        self.assertEqual(result.shape, result_shape)
+
+    def test_slicing_1d_normal(self):
+        # A 'normal' 1d testcase with no empty slices.
+        self._check_slicing(
+            test_shape=(3,),
+            indices=Slices[1:10],
+            result_shape=(2,),
+            data_was_fetched=True)
+
+    def test_slicing_1d_empty(self):
+        # A 1d testcase with an empty slicing.
+        self._check_slicing(
+            test_shape=(3,),
+            indices=Slices[0:0],
+            result_shape=(0,),
+            data_was_fetched=False)
+
+    def test_slicing_1d_unrecognised_empty(self):
```

Comment:
There are any number of cases that will return `result_shape=(0,)` and `data_was_fetched=True`, and I don't believe we get much value from just testing one of them. It would also be too much effort to try and cover all of them. This test should come out.

Reply:
Ok.